snuba/migrations/snuba_migrations/functions/0001_functions.py

```diff
@@ -0,0 +1,189 @@
+    Column("project_id", UInt(64)),
+    Column("transaction_name", String()),
+    Column("timestamp", DateTime()),
+    Column("depth", UInt(32)),
+    Column("parent_fingerprint", UInt(64)),
+    Column("fingerprint", UInt(64)),
+    Column("symbol", String()),
+    Column("image", String()),
+    Column("filename", String()),
+    Column("is_application", UInt(8)),
+    Column("tags", Nested([Column("key", String()), Column("value", String())])),
+    Column("retention_days", UInt(16)),
+]
+
+raw_columns: List[Column[Modifiers]] = common_columns + [
+    Column("durations", Array(Float(64))),
+    Column("profile_id", UUID()),
+]
+
+agg_columns: List[Column[Modifiers]] = common_columns + [
+    Column("count", AggregateFunction("count", [Float(64)])),
+    Column(
+        "percentiles",
+        AggregateFunction("quantiles(0.5, 0.75, 0.9, 0.95, 0.99)", [Float(64)]),
+    ),
+    Column("min", AggregateFunction("min", [Float(64)])),
+    Column("max", AggregateFunction("max", [Float(64)])),
+    Column("avg", AggregateFunction("avg", [Float(64)])),
+    Column("sum", AggregateFunction("sum", [Float(64)])),
+    Column("worst", AggregateFunction("argMax", [UUID(), Float(64)])),
+    Column("examples", AggregateFunction("groupUniqArray(5)", [UUID()])),
+]
+
+
+class Migration(migration.ClickhouseNodeMigration):
+    blocking = False
+    index_granularity = "2048"
+    storage_set = StorageSetKey.FUNCTIONS
+
+    data_granularity = 60 * 60
+
+    local_raw_table = "functions_raw_local"
+    dist_raw_table = "functions_raw_dist"
+
+    local_materialized_table = "functions_mv_local"
+    dist_materialized_table = "functions_mv_dist"
+
+    local_view_table = "functions_local"
+    dist_view_table = "functions_dist"
+
+    def forwards_local(self) -> Sequence[operations.SqlOperation]:
+        return [
+            operations.CreateTable(
+                storage_set=self.storage_set,
+                table_name=self.local_raw_table,
+                columns=raw_columns,
+                engine=table_engines.MergeTree(
+                    storage_set=self.storage_set,
+                    order_by="(org_id, project_id, transaction_name, timestamp)",
+                    partition_by="(toStartOfInterval(timestamp, INTERVAL 3 day))",
+                    ttl="timestamp + toIntervalDay(7)",
+                ),
+            ),
+            operations.CreateTable(
+                storage_set=self.storage_set,
+                table_name=self.local_materialized_table,
+                columns=agg_columns,
+                engine=table_engines.AggregatingMergeTree(
+                    storage_set=self.storage_set,
+                    order_by="(org_id, project_id, transaction_name, timestamp, depth, parent_fingerprint, fingerprint, symbol, image, filename, is_application, tags.key, tags.value, retention_days)",
+                    primary_key="(org_id, project_id, transaction_name, timestamp, depth, parent_fingerprint, fingerprint)",
```

Comment:
This matters when you filter your data. Do you expect to ever be filtering anything past timestamp? `depth, parent_fingerprint, fingerprint`? If not then keeping them here only makes the primary key less efficient as the primary index would not fit in memory.

Reply:
This will be useful for constructing visualizations flamegraphs/calltrees. In order to construct them, we need to page through the result frames ordered by depth, parent_fingeprint and fingerprint to be able to consistently connect frames to their parents.

The condition will look something like the following

```sql
SELECT
  symbol,
  image,
  filename,
  fingerprint,
  parent_fingerprint,
  count(),
  sum(duration)
FROM functions_dist
WHERE
  start_timestamp >= ...
  AND start_timestamp < ...
  AND project_id = ''
  AND transaction_name = ''
  AND ...
  // we need to order the frames by depth/parent_fingerprint/fingerprint
  // so we can page through the results and merge them together
  AND (
    (depth > max depth from last batch)
    OR (
      depth = max depth from last batch
      AND parent_fingerprint >= max parent fingerprint from max depth from last batch
      AND fingerprint >= max fingerprint from max depth from last batch
    )
GROUP BY symbol, address, fingerprint, parent_fingerprint
ORDER BY depth ASC, parent_fingerprint ASC, fingerprint ASC
LIMIT 1000
```

Would it make sense to remove these fields from the primary keys in this case?